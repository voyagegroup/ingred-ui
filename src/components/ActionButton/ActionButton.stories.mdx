import { Meta, Story, ArgsTable, Canvas } from "@storybook/addon-docs/blocks";
import ActionButton from "./ActionButton";
import { Flex, Spacer, Typography } from "..";

<Meta
  title="Components/Inputs/ActionButton"
  component={ActionButton}
  argTypes={{ onClick: { action: "clicked" } }}
/>

# ActionButton

It can behave like a `<button />` tags.

<ArgsTable of={ActionButton} />

## Samples

<Canvas>
  <Story name="Example">
    <Flex display="flex">
      <Spacer mr={5}>
        <Typography weight="bold">Primary</Typography>
        <Spacer pt={1} />
        <ActionButton icon="pencil" color="primary">
          Edit
        </ActionButton>
      </Spacer>
      <Spacer mr={5}>
        <Typography weight="bold">Warning</Typography>
        <Spacer pt={1} />
        <ActionButton icon="pencil" color="warning">
          Edit
        </ActionButton>
      </Spacer>
      <Spacer mr={5}>
        <Typography weight="bold">Disabled</Typography>
        <Spacer pt={1} />
        <ActionButton icon="pencil" disabled={true}>
          Edit
        </ActionButton>
      </Spacer>
    </Flex>
  </Story>
</Canvas>
