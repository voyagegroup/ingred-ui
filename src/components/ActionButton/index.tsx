export { default, ActionButtonProps } from "./ActionButton";
