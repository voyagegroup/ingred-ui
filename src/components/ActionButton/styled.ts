import styled from "styled-components";
import { Space } from "../../styles";

type ContainerProps = {
  normalBackgroundColor: string;
  hoverBackgroundColor: string;
};

export const Container = styled.button<ContainerProps>`
  display: flex;
  align-items: center;
  padding: ${Space * 0.75}px ${Space}px;
  border: 0;
  cursor: pointer;
  border-radius: ${({ theme }) => theme.radius}px;
  background-color: ${({ normalBackgroundColor }) => normalBackgroundColor};
  white-space: nowrap;
  transition: all 0.3s;

  &:hover {
    background-color: ${({ hoverBackgroundColor }) => hoverBackgroundColor};
  }

  &:disabled {
    cursor: not-allowed;
  }
`;
