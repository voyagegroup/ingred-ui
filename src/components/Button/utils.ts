import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
