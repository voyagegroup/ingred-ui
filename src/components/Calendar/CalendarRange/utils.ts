import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
