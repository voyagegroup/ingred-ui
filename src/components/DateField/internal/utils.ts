export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
