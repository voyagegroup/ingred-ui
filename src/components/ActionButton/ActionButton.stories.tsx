import { action } from "@storybook/addon-actions";
import { ArgsTable, Description, Stories, Title } from "@storybook/addon-docs";
import { Story } from "@storybook/react/types-6-0";
import React from "react";
import ActionButton, { ActionButtonProps } from "./ActionButton";

export default {
  title: "Components/Inputs/ActionButton",
  component: ActionButton,
  args: {
    onClick: action("clicked"),
    icon: "pencil",
    children: "Edit",
  },
  parameters: {
    docs: {
      source: { type: "code" },
      page: () => (
        <>
          <Title />
          <Description markdown={"It can behave like a `<button />` tags."} />
          <ArgsTable of={ActionButton} />
          <Stories includePrimary title="Stories" />
        </>
      ),
    },
  },
};

const Template: Story<ActionButtonProps> = (args) => <ActionButton {...args} />;

export const Primary = Template.bind({});

export const Warning = Template.bind({});
Warning.args = {
  color: "warning",
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};
