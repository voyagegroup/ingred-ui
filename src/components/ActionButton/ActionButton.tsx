import * as React from "react";
import * as Styled from "./styled";
import Spacer from "../Spacer";
import Icon from "../Icon";
import { IconName } from "../Icon/Icon";
import Typography from "../Typography";
import { Theme, useTheme } from "../../themes";
import { colors } from "../../styles/color";

const getNormalBackgroundColor = (theme: Theme) => ({
  primary: theme.palette.background.hint,
  warning: theme.palette.danger.highlight,
  disabled: theme.palette.gray.light,
});

const getHoverBackgroundColor = (theme: Theme) => ({
  primary: theme.palette.primary.highlight,
  warning: colors.red[200],
  disabled: theme.palette.gray.light,
});

const getTextColor = (theme: Theme) => ({
  primary: theme.palette.primary.main,
  warning: theme.palette.danger.main,
  disabled: theme.palette.text.disabled,
});

type ColorProp = "primary" | "warning";

export type ActionButtonProps = React.ComponentPropsWithoutRef<"button"> & {
  icon: IconName;
  color?: ColorProp;
};

const ActionButton = React.forwardRef<HTMLButtonElement, ActionButtonProps>(
  ({ children, icon, color = "primary", disabled, ...rest }, ref) => {
    const theme = useTheme();

    const getColorByDisabled = (
      color: ColorProp,
      disabled?: boolean,
    ): ColorProp | "disabled" => {
      return disabled ? "disabled" : color;
    };

    const colorForStyle = getColorByDisabled(color, disabled);
    const normalBackgroundColor =
      getNormalBackgroundColor(theme)[colorForStyle];
    const hoverBackgroundColor = getHoverBackgroundColor(theme)[colorForStyle];
    const textColor = getTextColor(theme)[colorForStyle];

    return (
      <Styled.Container
        {...rest}
        ref={ref}
        normalBackgroundColor={normalBackgroundColor}
        hoverBackgroundColor={hoverBackgroundColor}
        disabled={disabled}
      >
        <Spacer pr={0.25}>
          <Icon name={icon} color={textColor} />
        </Spacer>
        <Typography color={textColor} size="md">
          {children}
        </Typography>
      </Styled.Container>
    );
  },
);

export default ActionButton;
