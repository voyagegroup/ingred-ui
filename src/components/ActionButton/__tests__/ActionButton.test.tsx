import * as React from "react";
import "@testing-library/jest-dom";
import { cleanup } from "@testing-library/react";
import ActionButton from "../";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("ActionButton component testing", () => {
  afterEach(cleanup);

  test("ActionButton", () => {
    const { asFragment } = renderWithThemeProvider(
      <ActionButton icon="pencil" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
