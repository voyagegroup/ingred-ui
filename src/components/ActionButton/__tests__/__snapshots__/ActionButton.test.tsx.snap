// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButton component testing ActionButton 1`] = `
<DocumentFragment>
  <button
    class="sc-bdvvtL hTa-DxM"
  >
    <div
      class="sc-gsDKAQ hIeZob"
    >
      <div
        class="sc-dkPtRN gozTXZ"
        size="18"
      >
        <svg
          viewBox="0 0 18 18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0,0H18V18H0Z"
            fill="none"
          />
          <path
            d="M12.546,8.022,11.485,6.961,4.5,13.947v1.06H5.561l6.986-6.986Zm1.061-1.061L14.667,5.9,13.607,4.84,12.546,5.9ZM6.181,16.507H3V13.325L13.076,3.249a.75.75,0,0,1,1.06,0l2.122,2.122a.75.75,0,0,1,0,1.061L6.182,16.507Z"
            fill="#0B82F4"
            transform="translate(-0.75 -0.757)"
          />
        </svg>
      </div>
    </div>
    <p
      class="sc-hKwDye gxqPEs"
      color="#0B82F4"
      font-size="14px"
    />
  </button>
</DocumentFragment>
`;
